library/core/src/lib.rs

```diff
@@ -282,160 +284,50 @@
 #![feature(wasm_target_feature)]
 // tidy-alphabetical-end
 
+// We want to be able to `cargo miri test` this crate, but that's tricky:
+// This will build the crate itself again, so now we have two `core` crates, one in the sysroot and
+// one built as dependency of the test crate. That leads to duplicate lang item errors. `./x.py test
+// library` avoids this by making the sysroot exactly the crates that `cargo build` produces, but
+// Miri builds the sysroot itself so that does not work.
+// Instead, we use the `miri-test` feature to indicate that we're running `cargo miri test`, and
+// with that feature we turn this crate into a re-export of the sysroot crate. We only do this when
+// building the crate that will become a dependency, not when doing the actual (doc)test build.
+#[cfg(all(feature = "miri-test", not(any(test, doctest))))]
+extern crate core as realcore;
+#[cfg(all(feature = "miri-test", not(any(test, doctest))))]
+#[stable(feature = "miri_test", since = "1.0.0")]
+pub use realcore::*;
+
+// Otherwise, we build the crate as usual. To avoid having to repeat the same `cfg` so many times,
+// we `include!("lib_.rs")` with the main crate contents, and `cfg` the `include!`. However, some
+// macro-related things can't be behind the `include!` so we have to repeat the `cfg` for them.
+
 // allow using `core::` in intra-doc links
+#[cfg(not(all(feature = "miri-test", not(any(test, doctest)))))]
```

Comment:
This might be easier to read as:

```rust
#[cfg(any(not(feature = "miri-test"), test, doctest))]
```

Reply:
That's what I originally had, but I found it hard to read as one has to very carefully check where the `not` ends. Maybe like this?
```rust
#[cfg(any(test, doctest, not(feature = "miri-test")))]
```

I wish we could just use `&&`, `||`, `!` here...

Reply:
Also I think we should make the imports "obviously" negations of each other. So if this becomes `any(...)` then the `realcore` re-export should become `not(any(...))`.